src/CounterInterceptor/MyCounterInterceptor.cs

```diff
@@ -0,0 +1,137 @@
+namespace TemporalioSamples.CounterInterceptor;
+
+using System.Collections.Concurrent;
+using Temporalio.Activities;
+using Temporalio.Client;
+using Temporalio.Client.Interceptors;
+using Temporalio.Worker.Interceptors;
+using Temporalio.Workflows;
+
+public class MyCounterInterceptor : IClientInterceptor, IWorkerInterceptor
+{
+    private ConcurrentDictionary<string, Counts> counts = new();
+
+    public ConcurrentDictionary<string, Counts> Counts => counts;
+
+    public string WorkerInfo() =>
+        string.Join(
+                "\n",
+                counts.Select(kvp => $"** Workflow ID: {kvp.Key} {kvp.Value.WorkflowInfo()}"));
+
+    public string ClientInfo() =>
+        string.Join(
+                "\n",
+                counts.Select(kvp => $"** Workflow ID: {kvp.Key} {kvp.Value.ClientInfo()}"));
+
+    public ClientOutboundInterceptor InterceptClient(ClientOutboundInterceptor nextInterceptor) =>
+        new ClientOutbound(this, nextInterceptor);
+
+    public WorkflowInboundInterceptor InterceptWorkflow(WorkflowInboundInterceptor nextInterceptor) =>
+        new WorkflowInbound(this, nextInterceptor);
+
+    public ActivityInboundInterceptor InterceptActivity(ActivityInboundInterceptor nextInterceptor) =>
+        new ActivityInbound(this, nextInterceptor);
+
+    private sealed class ClientOutbound : ClientOutboundInterceptor
+    {
+        private MyCounterInterceptor root;
+
+        public ClientOutbound(MyCounterInterceptor root, ClientOutboundInterceptor next)
+            : base(next) => this.root = root;
+
+        public override Task<WorkflowHandle<TWorkflow, TResult>> StartWorkflowAsync<TWorkflow, TResult>(
+            StartWorkflowInput input)
+        {
+            var id = input.Options.Id ?? "None";
+            // Need to add an empty record if none exists
+            // we don't care if it doesn't add it as we will
+            // still increment the current value.
+            root.counts.TryAdd(id, new Counts());
```

Comment:
This should happen on every call, not just start/execute workflow. The `root.counts[id].ClientSignals` call below will throw an exception when someone uses this to signal a workflow that wasn't started here. Same for other situations. Consider just having a helper on the root like:

```csharp
private void Increment(string id, Func<Counts, ref uint> toIncrement) =>
    Interlocked.Increment(ref toIncrement(counts.GetOrAdd(id, _ => new()));
```

Or something like that (untested). Then you can just change this to `root.Increment(id ?? "None", c => ref c.ClientExecutions);`.

Reply:
I debated on whether it needed to be done on every call or not. 

I was able to use your example as you had it. Worked the first time. 

Thanks!